csrc/scheduler/normalization_inner.cpp

```diff
@@ -376,16 +376,15 @@ bool compareTwoHeuristics(
     return score > 0;
   }
 
-  // Large register overhead leaves more registers for the compiler to
-  // optimize generated SASS code and avoids register spills. But don't want
-  // to use a very large block size which increaes inter-thread communication.
-  // This condition influence dropout_layer_norm at 7K and softmax_dropout at
-  // 6K and 16K.
-  constexpr int64_t opt_max_threads_per_block = 512l;
+  // Ensure the count of non buffer registers is larger than the min overhead.
+  // But don't want to achieve this goal at the cost of using a very large block
+  // size, it avoids using a small persistent batch with a large block size,
+  // which usually leads to 10% lower in performance.
+  constexpr int64_t opt_max_threads_per_block = 512;
   score = compare(
```

Comment:
I tried to split this compare into two parts and tried different orders:
(1) block size after register count, orange color in the fig.
```
  score = compare(
      ha.non_buffer_registers > min_non_buffer_registers,
      hb.non_buffer_registers > min_non_buffer_registers);
  if (score != 0) {
    return score > 0;
  }
  score = compare(ha.padded_bdimx <= 512, hb.padded_bdimx <= 512);
  if (score != 0) {
    return score > 0;
  }

```
(2) block size before register count, green color in the fig.
```
  score = compare(ha.padded_bdimx <= 512, hb.padded_bdimx <= 512);
  if (score != 0) {
    return score > 0;
  }
  score = compare(
      ha.non_buffer_registers > min_non_buffer_registers,
      hb.non_buffer_registers > min_non_buffer_registers);
  if (score != 0) {
    return score > 0;
  }

```

At 22K, we have 3 candidates after filtered by target occupancy. 
(a) The original code won't have a preference for batch sizes 3,5,6 because no one has more than 16 non_buffer_register and also smaller than 512. So it move to next rule: pick the highest occupancy, and batch size 5 is selected.
(b) if separate the original rule into block size followed by reg count. It picks batch size 6 since it is the only one has `padded_bdimx <= 512`
(c) if put block size after reg count. It picks size 3 since it is the only one has `non_buffer_registers > min_non_buffer_registers`
Both (b) and (c) are slower than (a).


```
bdimx: 469, bdimy: 1, padded_bdimx: 480, persistent_batch_size: 6, register_per_thread: 64, non_buffer_registers: 16, occupancy: 30, n_persistent_tails: 4, is_pad_bdimx: 1
bdimx: 938, bdimy: 1, padded_bdimx: 960, persistent_batch_size: 3, register_per_thread: 64, non_buffer_registers: 40, occupancy: 30, n_persistent_tails: 1, is_pad_bdimx: 1
bdimx: 563, bdimy: 1, padded_bdimx: 576, persistent_batch_size: 5, register_per_thread: 56, non_buffer_registers: 16, occupancy: 36, n_persistent_tails: 4, is_pad_bdimx: 1
```

At 13K, the perfs are improved using  (b) or (c) because the occupancies are all very high, e.g. 39 if use (b), 36 if use (c), 52 if use (a).
```
bdimx: 416, bdimy: 1, padded_bdimx: 416, persistent_batch_size: 4, register_per_thread: 48, non_buffer_registers: 16, occupancy: 39, n_persistent_tails: 0, is_pad_bdimx: 1
bdimx: 832, bdimy: 1, padded_bdimx: 832, persistent_batch_size: 2, register_per_thread: 32, non_buffer_registers: 16, occupancy: 52, n_persistent_tails: 0, is_pad_bdimx: 1
bdimx: 332, bdimy: 1, padded_bdimx: 352, persistent_batch_size: 5, register_per_thread: 56, non_buffer_registers: 16, occupancy: 33, n_persistent_tails: 1, is_pad_bdimx: 1
bdimx: 554, bdimy: 1, padded_bdimx: 576, persistent_batch_size: 3, register_per_thread: 56, non_buffer_registers: 32, occupancy: 36, n_persistent_tails: 1, is_pad_bdimx: 1
```
![image](https://github.com/NVIDIA/Fuser/assets/116412316/8dc4afc1-8210-41b0-a853-5d086970f4eb)

Reply:
Thanks for the results. The differences just look too small to specialize the heuristics in such a fine-grained way. I think it's fine with the current code, but just make it clear that this heuristics is found experimentally.